### PR TITLE
Add OTLP translation

### PR DESCRIPTION
**Pull request by keitwb**
This adds the protocol/otlp package that can translate from OTLP metric types
to SignalFx datapoints.  It follows the same conversion process as the OTEL
collector, but that code converts to our protobuf format and not datapoint.Datapoint which is what we need for our ingest service.

### CODE REVIEW

File: protocol/otlp/metrics.go [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+				sfxDps = append(sfxDps, FromMetric(m)...)
+			}
+		}
+
+		extraDimensions := attributesToDimensions(rm.GetResource().GetAttributes())
+		for i := range sfxDps {
+			dpDims := sfxDps[i].Dimensions
+			for k, v := range extraDimensions {
+				if _, ok := dpDims[k]; !ok {
+					dpDims[k] = v
```
[the remark this answers is not in the document]
Comment (keitwb) [reply protocol/otlp/metrics.go:62]: Yes, I believe that is what is does by virtue of the `!ok` guard that will prevent the dim being set if it already is set in the datapoint.  Double check my logic though.

[the remark this answers is not in the document]
Comment (keitwb) [reply protocol/otlp/metrics.go:62]: I should probably just write a test for it though.

File: protocol/otlp/decoder.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+type httpMetricDecoder struct {
+	sink   dpsink.Sink
+	logger log.Logger
+	buffs  sync.Pool
+}
+
+func NewHTTPMetricDecoder(sink dpsink.Sink, logger log.Logger) signalfx.ErrorReader {
+	return &httpMetricDecoder{
+		sink:   sink,
+		logger: logger,
```
Comment (jgheewala) [inline protocol/otlp/decoder.go:26]: please use logging with context here

File: protocol/otlp/metrics.go [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+func FromOTLPMetricRequest(md *metricsservicev1.ExportMetricsServiceRequest) []*datapoint.Datapoint {
+	return FromOTLPResourceMetrics(md.GetResourceMetrics())
+}
+
+// FromMetrics converts OTLP ResourceMetrics to SignalFx datapoints.
+func FromOTLPResourceMetrics(rms []*metricsv1.ResourceMetrics) []*datapoint.Datapoint {
+	var sfxDps []*datapoint.Datapoint
+
+	for _, rm := range rms {
+		for _, ilm := range rm.GetInstrumentationLibraryMetrics() {
```
Comment (jgheewala) [inline protocol/otlp/metrics.go:51]: do we really care about this metrics ?

Comment (jgheewala) [reply protocol/otlp/metrics.go:51]: how is the conversion happening please share document

File: protocol/otlp/metrics.go [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+	}
+
+	data := m.GetData()
+	switch data.(type) {
+	case *metricsv1.Metric_Gauge:
+		dps = convertNumberDataPoints(m.GetGauge().GetDataPoints(), basePoint)
+	case *metricsv1.Metric_Sum:
+		dps = convertNumberDataPoints(m.GetSum().GetDataPoints(), basePoint)
+	case *metricsv1.Metric_Histogram:
+		dps = convertHistogram(m.GetHistogram().GetDataPoints(), basePoint)
```
Comment (jgheewala) [inline protocol/otlp/metrics.go:87]: does this mean we will support histo?

File: protocol/otlp/decoder.go (hunk 2 above)
Comment (keitwb) [reply protocol/otlp/decoder.go]: Done.

File: protocol/otlp/metrics.go (hunk 3 above)
Comment (jgheewala) [reply protocol/otlp/metrics.go:51]: so basically we convert it our model right ?

Comment (keitwb) [reply protocol/otlp/metrics.go:51]: yes, exactly.

**Review by jgheewala — APPROVED** [review]

**Pull request merged**